File: app/src/main/java/io/builderscon/conference2017/view/SessionsViewModel.java
package io.builderscon.conference2017.view;

import android.content.Context;
import android.databinding.BaseObservable;
import android.databinding.Bindable;
import android.support.annotation.NonNull;
import android.text.TextUtils;
import android.view.View;

import com.annimon.stream.Collectors;
import com.annimon.stream.Stream;

import java.util.ArrayList;
import java.util.Date;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import io.builderscon.client.model.Room;
import io.builderscon.client.model.Session;
import io.builderscon.client.model.Track;
import io.builderscon.conference2017.model.Timetable;
import io.builderscon.conference2017.model.repository.TimetableRepository;

public class SessionsViewModel extends BaseObservable {

    private TimetableRepository timetableRepository = new TimetableRepository();

    private List<Room> rooms;

    private List<Date> stimes;

    private Map<String, String> tracksMap;

    public List<SessionViewModel> getSessions(Context context, int position) {
        List<Timetable> timetables = timetableRepository.read();
        List<Session> sessions = timetables.get(position).getSessions();
        List<Track> tracks = timetables.get(position).getTracks();
        this.tracksMap = extractTracksMap(tracks);
        this.rooms = extractRooms(sessions);
        this.stimes = extractStimes(sessions);
        List<SessionViewModel> viewModels = Stream.of(sessions)
                .map(session -> new SessionViewModel(session, context))
                .toList();

        return adjustViewModels(viewModels);
    }

    private List<SessionViewModel> adjustViewModels(List<SessionViewModel> sessionViewModels) {
        // Prepare sessions map
        final Map<String, SessionViewModel> sessionMap = new LinkedHashMap<>();
        for (SessionViewModel viewModel : sessionViewModels) {
            String roomName = viewModel.getRoomName();
            if (TextUtils.isEmpty(roomName)) {
                // In the case of Welcome talk and lunch time, set dummy room
                roomName = rooms.get(0).getName();
            }
            sessionMap.put(generateStimeRoomKey(viewModel.getStime(), roomName), viewModel);
        }

        final List<SessionViewModel> adjustedViewModels = new ArrayList<>();

        for (Date stime : stimes) {
            final List<SessionViewModel> sameTimeViewModels = new ArrayList<>();
            int maxRowSpan = 1;
            for (int i = 0, size = rooms.size(); i < size; i++) {
                Room room = rooms.get(i);
                SessionViewModel viewModel = sessionMap.get(generateStimeRoomKey(stime, room.getName()));
                if (viewModel != null) {
                    sameTimeViewModels.add(viewModel);
                    if (viewModel.getRowSpan() > maxRowSpan) {
                        maxRowSpan = viewModel.getRowSpan();
                    }
                } else {
                    SessionViewModel empty = SessionViewModel.createEmpty(1);
                    sameTimeViewModels.add(empty);
                }
            }

            List<SessionViewModel> copiedTmpViewModels = new ArrayList<>(sameTimeViewModels);
            for (SessionViewModel tmpViewModel : sameTimeViewModels) {
                int rowSpan = tmpViewModel.getRowSpan();
                if (rowSpan < maxRowSpan) {
                    // Fill for empty cell
                    copiedTmpViewModels.add(SessionViewModel.createEmpty(maxRowSpan - rowSpan));
                }
            }

            adjustedViewModels.addAll(copiedTmpViewModels);
        }

        return adjustedViewModels;
    }

    private String generateStimeRoomKey(@NonNull Date stime, @NonNull String roomName) {
        return DateUtil.getLongFormatDate(stime) + "_" + roomName;
    }

    private Map<String, String> extractTracksMap(List<Track> tracks) {
        return Stream.of(tracks).collect(Collectors.toMap(Track::getRoomId, Track::getName));
    }

    private List<Date> extractStimes(List<Session> sessions) {
        return Stream.of(sessions)
                .map(Session::getStartsOn)
                .sorted()
                .distinct()
                .toList();
    }

    private List<Room> extractRooms(List<Session> sessions) {
        return Stream.of(sessions)
                .map(Session::getRoom)
                .filter(room -> room != null)
                .sorted((lhs, rhs) -> tracksMap.get(lhs.getId()).compareTo(tracksMap.get(rhs.getId())))
                .distinct()
                .toList();
    }

    public List<Room> getRooms() {
        return rooms;
    }

    public List<Date> getStimes() {
        return stimes;
    }

    public Map<String, String> getTracksMap() {
        return tracksMap;
    }


    @Bindable
    public int getLoadingVisibility() {
        if (this.rooms == null) {
            return View.VISIBLE;
        } else {
            return View.GONE;
        }
    }
}

File: app/src/main/kotlin/io/builderscon/conference2017/viewmodel/SessionsViewModel.kt
package io.builderscon.conference2017.viewmodel

import android.databinding.BaseObservable
import io.builderscon.client.model.Room
import io.builderscon.client.model.Session
import io.builderscon.client.model.Track
import io.builderscon.conference2017.extension.getLongFormatDate
import io.builderscon.conference2017.extension.ignoreSameTimes
import io.builderscon.conference2017.extension.ignoreTmpTimes
import io.builderscon.conference2017.extension.needAddEmptyTime
import io.builderscon.conference2017.model.repository.TimetableRepository
import java.util.*

class SessionsViewModel : BaseObservable() {

    private val timetableRepository = TimetableRepository()

    var rooms: List<Room> = emptyList()
    var sTimes: List<Date> = emptyList()
    var tracksMap: Map<String, String> = emptyMap()

    suspend fun getSessions(position: Int): List<SessionViewModel> {
        val timetables = timetableRepository.read()
        val sessions = timetables[position].sessions
        val tracks = timetables[position].tracks
        this.tracksMap = extractTracksMap(tracks)
        this.rooms = extractRooms(sessions)
        this.sTimes = extractSTimes(sessions)
        val viewModels = sessions.map { session -> SessionViewModel(session) }.toList()

        return adjustViewModels(viewModels)
    }

    private fun adjustViewModels(sessionViewModels: List<SessionViewModel>): List<SessionViewModel> {
        val sessionMap = LinkedHashMap<String, SessionViewModel>()
        for (viewModel in sessionViewModels) {
            val roomName = viewModel.roomName
            val mapKey = viewModel.getStime()?.let { generateSTimeRoomKey(it, roomName) }
            mapKey?.let { sessionMap.put(it, viewModel) }
        }

        val adjustedViewModels = ArrayList<SessionViewModel>()

        sTimes.forEach { sTime ->
            if (sTime.needAddEmptyTime()) {
                val empty = SessionViewModel.createEmpty(1)
                adjustedViewModels.add(empty)
            }
            val sameTimeViewModels = ArrayList<SessionViewModel>()
            var maxRowSpan = 1
            (0 until rooms.size).forEach { index ->
                val (_, _, name) = rooms[index]
                val viewModel = sessionMap[generateSTimeRoomKey(sTime, name)]
                viewModel?.let { vm ->
                    sameTimeViewModels.add(viewModel)
                    if (vm.rowSpan > maxRowSpan) {
                        maxRowSpan = vm.rowSpan
                    }
                } ?: run {
                    if (!sTime.ignoreSameTimes()) {
                        sameTimeViewModels.add(SessionViewModel.createEmpty(1))
                    }
                }
            }

            val copiedTmpViewModels = ArrayList(sameTimeViewModels)
            sameTimeViewModels.forEach {
                val rowSpan = it.rowSpan
                if (rowSpan < maxRowSpan && !sTime.ignoreTmpTimes()) {
                    // Fill for empty cell
                    copiedTmpViewModels.add(SessionViewModel.createEmpty(maxRowSpan - rowSpan))
                }
            }

            adjustedViewModels.addAll(copiedTmpViewModels)
        }
        return adjustedViewModels
    }

    private fun generateSTimeRoomKey(sTime: Date, roomName: String): String {
        return sTime.getLongFormatDate() + "_" + roomName
    }

    private fun extractTracksMap(tracks: List<Track>): Map<String, String> {
        return tracks.map { Pair(it.roomId, it.name) }.toMap()
    }

    private fun extractSTimes(sessions: List<Session>): List<Date> {
        return sessions.map { it.startsOn }.sorted().distinct()
    }

    private fun extractRooms(sessions: List<Session>): List<Room> {
        return sessions.map { it.room }
                .filterNotNull()
                .sortedBy { tracksMap[it.id] }
                .distinct()
    }
}
